File: daa.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <float.h>

typedef struct point {
    int x, y;
} point;

int comparex(const void *a, const void *b) {
    point *p1 = (point *)a, *p2 = (point *)b;
    return (p1->x - p2->x);
}

int comparey(const void *a, const void *b) {
    point *p1 = (point *)a, *p2 = (point *)b;
    return (p1->y - p1->y);
}

float dist(point p1, point p2) {
    return sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y));
}

float bruteforce(point ar[], int n) {
    float min = FLT_MAX;
    int i,j;
    for ( i = 0; i < n; i++) {
        for ( j = i + 1; j < n; j++) {
            if (dist(ar[i], ar[j]) < min) {
                min = dist(ar[i], ar[j]);
            }
        }
    }
    return min;
}

float strip_closest(point strip[], int j, float d) {
    float min = d;
    qsort(strip, j, sizeof(point), comparey);
    int i,m;
    for ( i = 0; i < j; i++) {
        for (m = i + 1; m < j && (strip[m].y - strip[i].y) < min; m++) {
            if (dist(strip[i], strip[m]) < min) {
                min = dist(strip[i], strip[m]);
            }
        }
    }
    return min;
}

float nearest_untill(point ar[], int n) {
    if (n <= 3) {
        return bruteforce(ar, n);
    }
    int mid = n / 2;
    point midpnt = ar[mid];
    float dl = nearest_untill(ar, mid);
    float dr = nearest_untill(ar + mid, n - mid);
    
    float d = (dl < dr) ? dl : dr;
    point strip[n];
    int j = 0,i;
    for ( i = 0; i < n; i++) {
        if (abs(ar[i].x - midpnt.x) < d) {
            strip[j++] = ar[i];
        }
    }
    return (d < strip_closest(strip, j, d)) ? d : strip_closest(strip, j, d);
}

float nearest(point ar[], int n) {
    qsort(ar, n, sizeof(point), comparex);
    return nearest_untill(ar, n);
}

int main() {
	int n,i;
	scanf("%d",&n);
    point ar[n];
    for(i=0;i<n;i++)
	{
		scanf("%d %d",&ar[i].x,&ar[i].y);
	}
    printf("The smallest distance between closest pair is: %f", nearest(ar, n));
    return 0;
}

File: daa.py
import math
import sys

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def compare_x(p1, p2):
    return p1.x - p2.x

def compare_y(p1, p2):
    return p1.y - p2.y

def dist(p1, p2):
    return math.sqrt((p1.x - p2.x)**2 + (p1.y - p2.y)**2)

def brute_force(arr, n):
    min_dist = sys.float_info.max
    for i in range(n):
        for j in range(i+1, n):
            if dist(arr[i], arr[j]) < min_dist:
                min_dist = dist(arr[i], arr[j])
    return min_dist

def strip_closest(strip, j, d):
    min_dist = d
    strip.sort(key=lambda x: x.y)
    for i in range(j):
        for m in range(i+1, j):
            if strip[m].y - strip[i].y < d:
                if dist(strip[i], strip[m]) < min_dist:
                    min_dist = dist(strip[i], strip[m])
            else:
                break
    return min_dist

def nearest_until(arr, n):
    if n <= 3:
        return brute_force(arr, n)
    
    mid = n // 2
    mid_point = arr[mid]
    
    dl = nearest_until(arr[:mid], mid)
    dr = nearest_until(arr[mid:], n-mid)
    
    d = min(dl, dr)
    
    strip = []
    j = 0
    for i in range(n):
        if abs(arr[i].x - mid_point.x) < d:
            strip.append(arr[i])
            j += 1
    
    return min(d, strip_closest(strip, j, d))

def nearest(arr, n):
    arr.sort(key=compare_x)
    return nearest_until(arr, n)

if __name__ == "__main__":
    n = int(input())
    arr = []
    for _ in range(n):
        x, y = map(int, input().split())
        arr.append(Point(x, y))
    
    print(f"The smallest distance between nearest pair is: {nearest(arr, n)}")


